modules/salesforce_mapping/src/Plugin/SalesforceMappingField/PropertiesExtended.php

```diff
@@ -0,0 +1,374 @@
+
+    // Resetting the `drupal_field_value` to just the `setting` portion,
+    // which should be a string.
+    $config_value = $form_state->getValue('config');
+    $config_value['drupal_field_value'] = $config_value['drupal_field_value']['setting'];
+    $form_state->setValue('config', $config_value);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function value(EntityInterface $entity, SalesforceMappingInterface $mapping) {
+    // No error checking here. If a property is not defined, it's a
+    // configuration bug that needs to be solved elsewhere.
+    // Multipicklist is the only target type that handles multi-valued fields.
+    $describe = $this
+      ->salesforceClient
+      ->objectDescribe($mapping->getSalesforceObjectType());
+    $field_definition = $describe->getField($this->config('salesforce_field'));
+    if ($field_definition['type'] == 'multipicklist') {
+      $values = [];
+      foreach ($entity->get($this->config('drupal_field_value')) as $value) {
+        $values[] = $this->getStringValue($entity, $value);
+      }
+      return implode(';', $values);
+    }
+    else {
+      return $this->getStringValue($entity, $this->config('drupal_field_value'));
+    }
+  }
+
+  /**
+   * Data Fetcher wrapper.
+   *
+   * @return \Drupal\typed_data\DataFetcherInterface
+   *   Data fetcher service.
+   *
+   * @throws \Exception
+   *   If typed_data.data_fetcher service does not exist.
+   */
+  protected function getDataFetcher() {
+    if (!\Drupal::hasService('typed_data.data_fetcher')) {
+      throw new \Exception('Module typed_data must be installed to use Extended Properties');
+    }
+    if (empty($this->dataFetcher)) {
+      $this->dataFetcher = \Drupal::service('typed_data.data_fetcher');
+    }
+    return $this->dataFetcher;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pullValue(SObject $sf_object, EntityInterface $entity, SalesforceMappingInterface $mapping) {
+    $field_selector = $this->config('drupal_field_value');
+    $pullValue = parent::pullValue($sf_object, $entity, $mapping);
+    try {
+      // Fetch the TypedData property and set its value.
+      $data = $this->getDataFetcher()->fetchDataByPropertyPath($entity->getTypedData(), $field_selector);
+      $data->setValue($pullValue);
+      return $data;
+    }
+    catch (MissingDataException $e) {
+
+    }
+    catch (\Drupal\typed_data\Exception\InvalidArgumentException $e) {
+
+    }
+    // Allow any other exception types to percolate.
+    // If the entity doesn't have any value in the field, data fetch will
+    // throw an exception. We must attempt to create the field.
+    // Typed Data API doesn't provide any good way to initialize a field value
+    // given a selector. Instead we have to do it ourselves.
+    // We descend only to the first-level fields on the entity. Cascading pull
+    // values to entity references is not supported.
+    $parts = explode('.', $field_selector, 4);
+
+    switch (count($parts)) {
```

Comment:
This looks rough -- is this the piece you were referring to in https://www.drupal.org/project/salesforce/issues/2899460#comment-12872244?

Reply:
Yeah it's awful.
I still haven't found a better way to do it though.
Typed Data and property selectors seem to be geared exclusively about fetching data, not setting it.

see https://www.drupal.org/project/typed_data/issues/3014934